Tests/RegexTests/MatchTests.swift

```diff
@@ -0,0 +1,403 @@
+import XCTest
+@testable import _MatchingEngine
+@testable import _StringProcessing
+import Algorithms
+
+func matchTest(
+  _ input: String, _ expecting: AST,
+  syntax: SyntaxOptions = .traditional
+) {
+  XCTExpectFailure {
+    XCTFail()
+  }
+}
+
+func matchTest(
+  _ regex: String,
+  input: String,
+  match: String,
+  syntax: SyntaxOptions = .traditional,
+  xfail: Bool = false
+) {
+  guard !xfail else {
+    XCTExpectFailure {
+      XCTFail()
+    }
+    return
+  }
+
+  let pattern = Algorithms.Regex(regex)
+  let range = input.firstRange(of: pattern)!
+  XCTAssertEqual(String(input[range]), match)
+}
+
+extension RegexTests {
+  func testMatch() {
+    matchTest(
+      "abc", input: "123abcxyz", match: "abc")
+    matchTest(
+      #"abc\+d*"#, input: "123abc+xyz", match: "abc+")
+    matchTest(
+      #"abc\+d*"#, input: "123abc+dddxyz", match: "abc+ddd")
+    matchTest(
+      "a(b)", input: "123abcxyz", match: "ab")
+    matchTest(
+      "abc(?:de)+fghi*k|j", input: "123abcdefghijxyz", match: "j")
+    matchTest(
+      "abc(?:de)+fghi*k|j", input: "123abcdedefghkxyz", match: "abcdedefghk")
+    matchTest(
+      "a(?:b|c)?d", input: "123adxyz", match: "ad")
+    matchTest(
+      "a(?:b|c)?d", input: "123abdxyz", match: "abd")
+    matchTest(
+      "a(?:b|c)?d", input: "123acdxyz", match: "acd")
+    matchTest(
+      "a?b??c+d+?e*f*?", input: "123abcdefxyz", match: "abcde")
+    matchTest(
+      "a?b??c+d+?e*f*?", input: "123bcddefxyz", match: "bcd")
+    matchTest(
+      "a|b?c", input: "123axyz", match: "a")
+    matchTest(
+      "a|b?c", input: "123bcxyz", match: "bc")
+    matchTest(
+      "(a|b)c", input: "123abcxyz", match: "bc")
+    matchTest(
+      "(.)*(.*)", input: "123abcxyz", match: "123abcxy", xfail: true)
+    matchTest(
+      #"abc\d"#, input: "xyzabc123", match: "abc1")
+
+    // MARK: Unicode scalars
+
```

Comment:
I don't think it makes sense to duplicate parse tests below, but it would be nice to convert the below into end-to-end tests.